addons/l10n_jo_edi/models/account_edi_xml_ubl_21_jo.py

```diff
@@ -0,0 +1,302 @@
+        if invoice._get_jo_invoice_type_number() != 4:
+            vals[0]['tax_subtotal_vals'] = []
+        return vals
+
+    def _get_invoice_line_item_vals(self, line, taxes_vals):
+        product = line.product_id
+        description = line.name and line.name.replace('\n', ', ')
+        return {
+            'name': product.name or description,
+        }
+
+    def _get_document_allowance_charge_vals_list(self, invoice):
+        return [{
+            'charge_indicator': 'false',
+            'allowance_charge_reason': 'discount',
+            'currency_name': JO_CURRENCY.name,
+            'currency_dp': self._get_currency_decimal_places(),
+            'amount': invoice._get_discount_amount_jod(),
+        }]
+
+    def _get_invoice_line_allowance_vals_list(self, line, tax_values_list=None):
+        return [{
+            'charge_indicator': 'false',
+            'allowance_charge_reason': 'DISCOUNT',
+            'currency_name': JO_CURRENCY.name,
+            'currency_dp': self._get_currency_decimal_places(),
+            'amount': line._get_discount_amount_jod(),
+        }]
+
+    def _get_invoice_line_price_vals(self, line):
+        vals = super()._get_invoice_line_price_vals(line)
+        return {
+            **vals,
+            'price_amount': line._get_unit_price_jod(),
+            'product_price_dp': self._get_currency_decimal_places(),
+            'allowance_charge_vals': self._get_invoice_line_allowance_vals_list(line),
+            'currency': JO_CURRENCY,
+        }
+
+    def _get_invoice_line_tax_totals_vals_list(self, line, taxes_vals):
+        if line.move_id._get_jo_invoice_type_number() in [1, 2]:
+            return []
+
+        invoice = line.move_id
+        vals = self._get_invoice_tax_totals_vals_list_helper(invoice, taxes_vals)
+        val = vals[0]
+        val['rounding_amount'] = line._get_tax_inclusive_amount_jod()
+
+        for grouping_key, tax_details_vals in taxes_vals['tax_details'].items():
+            if grouping_key['tax_amount_type'] == 'fixed':
+                subtotal = {
+                    'currency': JO_CURRENCY,
+                    'currency_dp': self._get_currency_decimal_places(),
+                    'taxable_amount': sum(record._get_tax_exclusive_amount_jod() - record._get_discount_amount_jod() for record in tax_details_vals['records']),
+                    'tax_amount': sum(record._get_special_tax_amount_jod() for record in tax_details_vals['records']),
+                    'tax_category_vals': tax_details_vals['_tax_category_vals_'],
+                }
+                val['tax_subtotal_vals'].insert(0, subtotal)
+        return vals
+
+    def _get_invoice_line_vals(self, line, line_id, taxes_vals):
+        if line.quantity < 0:
+            raise ValidationError("JoFotara portal cannot process negative quantity on invoice line")
+        if line.price_unit < 0:
+            raise ValidationError("JoFotara portal cannot process negative price on invoice line")
```

Comment:
Should I add a constraint on the `quantity` and `price_unit` preventing them from being negative for Jordanian companies? I thought about this at the beginning, but I remember being told a counter argument that we needn't always prevent it, and we only need to prevent it at the XML generation time. If you think the fields constraints make more sense, I would go for it.